tls/s2n_handshake_io.c

```diff
@@ -1040,27 +1015,44 @@ int s2n_negotiate(struct s2n_connection *conn, s2n_blocked_status * blocked)
         this = 'C';
     }
 
-    while (ACTIVE_STATE(conn).writer != 'B') {
-        /* Flush any pending I/O or alert messages */
-        GUARD(s2n_flush(conn, blocked));
+    /* If we were blocked reading a record, then the handler is waiting on external data.
+     * The handler will know how to continue, so we should call the handler right away.
+     * We aren't going to read more handshake data yet because the current message has
+     * not finished processing. */
+    if (conn->handshake.paused) {
```

Comment:
Might be worth moving some/all this logic into another function, the way handshake_read_io and handshake_write_io are. Call it "handshake_retry_state" or something?